satellite/metabase/rangedloop/service.go

```diff
@@ -74,81 +86,116 @@ func (service *Service) Run(ctx context.Context) (err error) {
-		}
-
-		// Create closure to capture loop variables.
-		createClosure := func(ctx context.Context, rangeIndex int, rangeProvider SegmentProvider, rangeObservers []Partial) func() error {
-			return func() (err error) {
-				defer mon.Task()(&ctx)(&err)
-
-				return rangeProvider.Iterate(ctx, func(segments []segmentloop.Segment) error {
-					for _, rangeObserver := range rangeObservers {
-						err := rangeObserver.Process(ctx, segments)
-						if err != nil {
-							return err
-						}
-					}
-					return nil
-				})
+			rangeState := rangeObserverState{
+				rangeObserver: rangeObserver,
+				duration:      new(time.Duration),
 			}
+			rangeObservers = append(rangeObservers, rangeState)
+			observerStates[i].rangeObservers = append(observerStates[i].rangeObservers, rangeState)
 		}
 
-		group.Go(createClosure(ctx, rangeIndex, rangeProvider, rangeObservers))
+		// Create closure to capture loop variables.
+		group.Go(createGoroutineClosure(ctx, rangeProvider, rangeObservers))
 	}
 
 	// Improvement: stop all ranges when one has an error.
 	errList := group.Wait()
 	if errList != nil {
-		return errs.Combine(errList...)
+		return nil, errs.Combine(errList...)
 	}
 
-	// Segment loop has ended.
-	// This is the reduce step.
-	for _, state := range observerStates {
-		for _, rangeObserver := range state.rangeObservers {
-			err := state.observer.Join(ctx, rangeObserver)
-			if err != nil {
-				return err
+	return finishObservers(ctx, observerStates)
+}
+
+func createGoroutineClosure(ctx context.Context, rangeProvider SegmentProvider, states []rangeObserverState) func() error {
+	return func() (err error) {
+		defer mon.Task()(&ctx)(&err)
+
+		return rangeProvider.Iterate(ctx, func(segments []segmentloop.Segment) error {
+			for _, state := range states {
+				start := time.Now()
+				err := state.rangeObserver.Process(ctx, segments)
+				if err != nil {
+					return err
+				}
+				*state.duration += time.Since(start)
 			}
+			return nil
+		})
+	}
+}
+
+func startObservers(ctx context.Context, observers []Observer) (observerStates []observerState, err error) {
+	startTime := time.Now()
+
+	for _, obs := range observers {
+		state, err := startObserver(ctx, startTime, obs)
+		if err != nil {
+			return nil, err
+		}
+
+		observerStates = append(observerStates, state)
+	}
+
+	return observerStates, nil
+}
+
+func startObserver(ctx context.Context, startTime time.Time, observer Observer) (observerState, error) {
+	err := observer.Start(ctx, startTime)
+
+	return observerState{
+		observer: observer,
+	}, err
+}
+
+func finishObservers(ctx context.Context, observerStates []observerState) (observerDurations []ObserverDuration, err error) {
+	for _, state := range observerStates {
+		observerDuration, err := finishObserver(ctx, state)
+		if err != nil {
+			return nil, err
 		}
 
-		err := state.observer.Finish(ctx)
+		observerDurations = append(observerDurations, observerDuration)
+	}
+
+	return observerDurations, err
```

Comment:
Returning nil explicitly is more self documenting that at this point the function has succeeded. When I see `err` here, I start to doubt that I understood the function correctly and scan back to see if I can find where it might have been set to something other than nil.
```suggestion
	return observerDurations, nil
```

Reply:
Done.